### PR TITLE
C++: #pragma once and a bugfix

### PR DESCRIPTION
**Pull request by lynn**
* Generate `#pragma once` in C++ code.
* Generate correct C++ code for decoding into a message field inside a message.

### CODE REVIEW

**Review by AndrewMD5 — APPROVED** [review]

**Review by apecoraro — COMMENTED** [review]
At first glance, albeit very quick, I can't tell if there is anything in this PR that addresses this issue:
```C++
static void decodeInto(::bebop::Reader& reader, AvailableCodecRequest& target) {
    const auto length = reader.readMessageLength();
    const auto end = reader.pointer() + length;
    while (true) {
      switch (reader.readByte()) {
        case 0:
          return;
        case 1:
          target.params = AvailableCodecRequestParams();
          AvailableCodecRequestParams::decodeInto(reader, target.params.value());
          break;
        default:
          reader.seek(end);
          return;
      }
    }
  }
```
Specifically the need to create a value on a struct before attempting to write to it, i.e. this line: `target.params = AvailableCodecRequestParams();` and then also to pass the reference to that value rather than a reference to the optional<type>, i.e. this line: `AvailableCodecRequestParams::decodeInto(reader, target.params.value());`

So just wanted to explicitly ask if that is fixed here?

File: Core/Generators/CPlusPlus/CPlusPlusGenerator.cs [hunk 1]
```diff
@@ -271,6 +271,7 @@ private string CompileDecodeField(TypeBase type, string target, int depth = 0, i
                 },
                 DefinedType dt when Schema.Definitions[dt.Name] is EnumDefinition =>
                     $"{target} = static_cast<{dt.Name}>(reader.readUint32());",
+                DefinedType dt when isOptional => $"{target}.emplace({dt.Name}::decode(reader));",
```
Comment (lynn) [inline Core/Generators/CPlusPlus/CPlusPlusGenerator.cs:274]: The fix is this line @apecoraro. Calling [`emplace`](https://en.cppreference.com/w/cpp/utility/optional/emplace) on the optional value achieves a similar thing to initializing it and then writing into `.value()` but it is somewhat more efficient, at least as far as I understand 🙂 .

Comment (lynn) [reply Core/Generators/CPlusPlus/CPlusPlusGenerator.cs:274]: So the generated code now looks like 

```cpp
        case 1:
          target.params.emplace(AvailableCodecRequestParams::decode(reader));
          break;
```

Comment (apecoraro) [reply Core/Generators/CPlusPlus/CPlusPlusGenerator.cs:274]: Nice!

**Pull request merged**